File: python/redis_work_queue/item.py
import uuid
import json


class Item(dict):
    """An item for a work queue. Each item has an ID and associated data."""

    def __init__(self, data: bytes | str, id=None):
        """
        Args:
            data (bytes or str): Data to associate with this item, strings will be converted to
                                 bytes.
            id (str | None): ID of the Item, if None, a new (random UUID) ID is generated.
        """
        if isinstance(data, str):
            data = bytes(data, 'utf-8')
        elif not isinstance(data, bytes):
            data = bytes(data)

        if id is None:
            id = uuid.uuid4().hex
        elif type(id) != str:
            id = str(id)

        dict.__init__(self, data=data, id=id)

    @classmethod
    def from_dict(cls, loaded: dict):
        """Create an `Item` from a dictionary containing 'data' and, optionally, 'id'."""
        id = None
        if 'id' in loaded:
            id = loaded['id']
        return cls(loaded['data'], id=id)

    @classmethod
    def parse(cls, s: str):
        """Parse an `Item` from JSON. The JSON structure should be an object with a 'data' key and,
        optionally, an 'id' key."""
        return cls.from_dict(json.loads(s))

    @classmethod
    def from_json_data(cls, data, id=None):
        """Generate an item where the associated data is the JSON string of `data`."""
        return cls(json.dumps(data, separators=(',', ':')), id=id)

    def data(self) -> bytes:
        """Get the data associated with this item."""
        return self['data']

    def data_json(self):
        """Get the data associated with this item, parsed as JSON."""
        return json.loads(self.data())

    def id(self) -> str:
        """Get the ID of the item."""
        return self['id']

File: typescript/redis_work_queue/Item.ts
const { v4: uuidv4 } = require('uuid');


type ItemData = {
  [key: string]: any;
};

export class Item {
  // An item for a work queue. Each item has an ID and associated data.

  public data: Buffer | string;
  public id: string;

  constructor(data: string | Buffer, id?: string) {
    /**
     * Args:
     * data (bytes or str): Data to associate with this item, strings will be converted to bytes.
     * id (str | null): ID of the Item, if null, a new (random UUID) ID is generated.
     */
    if (!(data instanceof Buffer)) {
      this.data = Buffer.from(data);
    } else {
      this.data = data;
    }

    if (id == null) {
      this.id = uuidv4();
    } else if (typeof id !== 'string') {
      this.id = String(id);
    } else {
      this.id = id;
    }

  
  }

  static fromDict(loaded: ItemData): Item {
    // Create an `Item` from a dictionary containing 'data' and, optionally, 'id'.
    let id: string | undefined = undefined;
    if ('id' in loaded) {
      id = loaded['id'];
    }
    return new Item(loaded['data'], id);
  }

  static parse(string: string) {
    // Parse an `Item` from JSON. The JSON structure should be an object with a 'data' key and, optionally, an 'id' key.
    return JSON.parse(string);
  }

  static fromJsonData(data: string, id?: string) {
    //Generate an item where the associated data is the JSON string of `data`.
    return new Item(JSON.stringify(data), id);
  }

  Data() {
    // Get the data associated with this item.
    return this.data;
  }

  dataJson(): any {
    // Get the data associated with this item, parsed as JSON.
    let jsonString: string;
    if (Buffer.isBuffer(this.data)) {
      jsonString = this.data.toString('utf-8');
    } else {
      jsonString = this.data;
    }
    try {
        console.log(jsonString)
      return JSON.parse(jsonString);
    } catch (error) {
      return {};
    }
  }
  
  

  Id(): string {
    // Get the ID of the item.
    return this.id;
  }
}
